firmware/hackrf_usb/sgpio_m0.s

```diff
@@ -1,108 +1,673 @@
+
+	// The worst case timing is assumed to occur when reading the interrupt
+	// status register *just* misses the flag being set - so we include the
+	// cycles required to check it a second time.
+	//
+	// We also assume that we can spend a full 10 cycles doing an ldr from
+	// SGPIO the first time (2 for ldr, plus 8 for SGPIO-AHB bus latency),
+	// and still miss a flag that was set at the start of those 10 cycles.
+	//
+	// This latter asssumption is probably slightly pessimistic, since the
+	// sampling of the flag on the SGPIO side must occur some time after
+	// the ldr instruction begins executing on the M0. However, we avoid
+	// relying on any assumptions about the timing details of a read over
+	// the SGPIO to AHB bridge.
+
+\name\()_int_wait:
+	// Spin on the exchange interrupt status, shifting the slice A flag to the carry flag.
+	ldr int_status, [sgpio_int, #INT_STATUS]        // int_status = SGPIO_STATUS_1          // 10, twice
+	lsr scratch, int_status, #1                     // scratch = int_status >> 1            // 1, twice
+	bcc \name\()_int_wait                           // if !carry: goto int_wait             // 3, then 1
+
+	// Clear the interrupt pending bits that were set.
+	str int_status, [sgpio_int, #INT_CLEAR]         // SGPIO_CLR_STATUS_1 = int_status      // 8
+.endm
+
+.macro update_buf_ptr
+	// Update the address of the buffer segment we want to write to / read from.
+	mov buf_ptr, buf_mask                           // buf_ptr = buf_mask                   // 1
+	and buf_ptr, count                              // buf_ptr &= count                     // 1
+	add buf_ptr, buf_base                           // buf_ptr += buf_base                  // 1
+.endm
+
+.macro update_counts
+	// Update counts after successful SGPIO operation.
+
+	// Update the byte count and store the new value.
+	add count, #32                                  // count += 32                          // 1
+	str count, [state, #M0_COUNT]                   // state.m0_count = count               // 2
+
+	// We didn't have a shortfall, so the current shortfall length is zero.
+	mov shortfall_length, hi_zero                   // shortfall_length = hi_zero           // 1
+.endm
+
+.macro jump_next_mode name
+	// Jump to next mode if the byte count threshold has been reached.
+	//
+	// Clobbers:
+	threshold .req r0
+	new_mode .req r1
+
+	// Check count against threshold. If not a match, return to start of current loop.
+	ldr threshold, [state, #THRESHOLD]              // threshold = state.threshold          // 2
+	cmp count, threshold                            // if count != threshold:               // 1
+	bne \name\()_loop                               //      goto loop                       // 1 thru, 3 taken
+
+	// Otherwise, load and set new mode.
+	ldr new_mode, [state, #NEXT_MODE]               // new_mode = state.next_mode           // 2
+	str new_mode, [state, #MODE]                    // state.mode = new_mode                // 2
```

Comment:
I think there is a very low probability race condition where the M4 sets the mode, but then the M0 clobbers it immediately with the contents of NEXT_MODE.  This could be mitigated by having the M4 set NEXT_MODE before setting MODE; however, that could result in a different problem where the M0 follows the new NEXT_MODE in `jump_next_mode` before checking to see if the M4 has changed MODE.  In this case the new mode would be entered without resetting the stats.